File: Leetcode/236/236.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */


class Solution2 {
public:
    unordered_map<int, TreeNode*> val2node;
    unordered_map<int, int> father;
    void dfs_1(TreeNode* current) {
        if (current == nullptr) return;
        val2node[current->val] = current;
        dfs_1(current->left);
        dfs_1(current->right);
    }
    void dfs_2(TreeNode* parent, TreeNode* son) {
        if (son == nullptr) return;
        if (parent) {
            father[son->val] = parent->val;
        }
        dfs_2(son, son->left);
        dfs_2(son, son->right);
    }
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        dfs_1(root);
        dfs_2(nullptr, root);
        unordered_set<int> vis;
        TreeNode* pCopy = p;
        while (true) {
            vis.insert(pCopy->val);
            if (pCopy == root) break;
            pCopy = val2node[father[pCopy->val]];
        }
        TreeNode* qCopy = q;
        while (true) {
            if (vis.find(qCopy->val) != vis.end()) {
                return qCopy;
            }
            qCopy = val2node[father[qCopy->val]];
        }
        return nullptr;
    }
};

class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        if (!root) return nullptr;
        if (root == p) return p;
        if (root == q) return q;
        TreeNode* lcaOfLeftSubtree = lowestCommonAncestor(root->left, p, q);
        TreeNode* lcaOfRightSubtree = lowestCommonAncestor(root->right, p, q);
        if (lcaOfLeftSubtree && lcaOfRightSubtree) {
            return root;
        }
        return lcaOfLeftSubtree ? lcaOfLeftSubtree : lcaOfRightSubtree;
    }
};

File: Leetcode/236/236.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func dfs1(node *TreeNode, n *int) {
	if node != nil {
		*n++
		dfs1(node.Left, n)
		dfs1(node.Right, n)
	}
}

func dfs2(node *TreeNode, node2id *map[int]int, id *int) {
	if node != nil {
		(*node2id)[node.Val] = *id
		*id++
		dfs2(node.Left, node2id, id)
		dfs2(node.Right, node2id, id)
	}
}

func dfs3(node *TreeNode, fa *TreeNode, father []int, node2id *map[int]int) {
	if node != nil {
		if fa != nil {
			father[(*node2id)[node.Val]] = (*node2id)[fa.Val]
		} else {
			father[(*node2id)[node.Val]] = -1
		}
		dfs3(node.Left, node, father, node2id)
		dfs3(node.Right, node, father, node2id)
	}
}

var (
	result *TreeNode
)

func dfs4(root *TreeNode, node2id *map[int]int, targetId int) {
	if result != nil {
		return
	}
	if root != nil {
		if (*node2id)[root.Val] == targetId {
			result = root
			return
		}
		dfs4(root.Left, node2id, targetId)
		dfs4(root.Right, node2id, targetId)
	}
}

func lowestCommonAncestor(root, p, q *TreeNode) *TreeNode {
	n := 0
	dfs1(root, &n)
	father := make([]int, n)
	node2id := make(map[int]int)
	id := 0
	dfs2(root, &node2id, &id)
	dfs3(root, nil, father, &node2id)
	pId := node2id[p.Val]
	unmap := make(map[int]struct{})
	for pId != -1 {
		unmap[pId] = struct{}{}
		pId = father[pId]
	}
	qId := node2id[q.Val]
	ans := -1
	for qId != -1 {
		if _, ok := unmap[qId]; ok {
			ans = qId
			break
		}
		qId = father[qId]
	}
	if ans == -1 {
		return nil
	}
	result = nil
	dfs4(root, &node2id, ans)
	return result
}